### PR TITLE
using Zygote

### PR DESCRIPTION
**Pull request by MikeInnes**
Otherwise known as "break all the things". This will be a huge change so I'm beginning to prepare now, even though Zygote is still a couple of months off from being really ready. **Do not try this at home** (yet) – this branch is eventually aimed at beta testers, but isn't even ready for that yet.

The idea is to break as little code as possible, which means supporting the current `Params` API; but I also want to start prototyping the nicer things discussed in #628 and other issues.

Blocking issues:

* [x] Get the tests passing.
* [x] Check tests on GPU.
* [x] Rewrite all the docs.
* [x] Cache invalidation (https://github.com/jrevels/Cassette.jl/issues/6).
* [x] Moving over adjoints (https://github.com/FluxML/Zygote.jl/issues/81).
* [x] General Zygote robustness.

Nice to have:

* [ ] Robust nested AD (may not be a blocker if one can still use Tracker with Flux).
* [x] Zygote support for modules / globals as discussed in #628, along with #637.
* [x] Better train/test mode as in #643.

If you're the kind of person who ignores triangular road signs, you can try this with

```julia
]add Flux#zygote Zygote#master
```

### CODE REVIEW

**Comment by MikeInnes** [comment]
I initially had started doing #666 and #637 on this branch, but that's turning into a big project, so I've stripped the commits for now (preserved on [mji/step](https://github.com/FluxML/Flux.jl/tree/mji/step)). As excited as I am to get rid of `Params` I think the right move is to shelve that for now and focus on the core AD issues.

**Comment by staticfloat** [comment]
In my own experiments trying to use `Zygote` with `Flux` (I just do `model = mapleaves(Flux.data, model)` first, then define my own Zygote-based update step:

```julia
zyg_update!(opt, model, updates::Nothing) = nothing
function zyg_update!(opt, model::AbstractArray, updates::AbstractArray)
    # Sub off to Flux's ADAM optimizer
    Δ = Flux.Optimise.update!(opt, model, updates)
    return model .-= Δ
end

function zyg_update!(opt, model, updates)
    if nfields(model) == 0
        return model
    end

    for field_idx in 1:nfields(model)
        zyg_update!(opt, getfield(model, field_idx), getfield(updates, field_idx))
    end
end
```

Things actually work fairly well, except BatchNorm freaks out, complaining about mutating arrays.  To work around this, I am using [my own `BatchNorm` implementation](https://gist.github.com/staticfloat/a509b1e1cb1fb556028779722c2531e6), re-architected to work with Zygote.  Not sure if that is the direction you want to go with this Mike, but it worked well for us on TPU.  I will note, anecdotally, that my convolution and batchnom-heavy workload (a convolutional autoencoder for large images) uses ~20% less memory with Zygote than with Tracker.

**Comment by staticfloat** [comment]
I wanted to use this with the new NNlib overhaul, so I rebased this branch on top of `master`; I'm not certain I did everything right, but `sf/zygote_updated` contains my rebased version.  Mike, if you like it, you can just force-push it to `#zygote` and keep on working, or you can just do the rebase yourself.

File: test/cuda/cuda.jl [hunk 1]
```diff
@@ -6,8 +6,6 @@ using Zygote
 
 @testset "CuArrays" begin
 
-CuArrays.allowscalar(false)
```
Comment (MikeInnes) [inline test/cuda/cuda.jl]: This is there for a reason. Why not fix the failing test?

File: test/cuda/cudnn.jl [hunk 2]
```diff
@@ -1,48 +1,47 @@
...
 
         @test cpu(data(cy)) ≈ data(y)
 
         g = rand(size(y)...)
-        Flux.back!(y, g)
-        Flux.back!(cy, gpu(g))
+        # Flux.back!(y, g)
+        # Flux.back!(cy, gpu(g))
 
-        @test m.γ.grad ≈ cpu(cm.γ.grad)
```
Comment (MikeInnes) [inline test/cuda/cudnn.jl]: @thebhatman you may already be aware, but any tests that use `back!` and `grad` like this will have to be changed to get a gradient using Zygote and test that.

File: test/layers/normalisation.jl [hunk 3]
```diff
@@ -1,29 +1,28 @@
...
+
+trainmode(f, x...) = forward(f, x...)[1]
 
 @testset "Dropout" begin
   x = [1.,2.,3.]
-  @test x == testmode!(Dropout(0.1))(x)
-  @test x == Dropout(0)(x)
-  @test zero(x) == Dropout(1)(x)
+  @test x == Dropout(0.1)(x)
+  @test x == trainmode(Dropout(0), (x))
```
Comment (MikeInnes) [inline test/layers/normalisation.jl]: Some redundant brackets have slipped in here

File: test/cuda/cuda.jl [hunk 4]
```diff
@@ -14,7 +16,7 @@ cx = gpu(x)
 
 x = Flux.onehotbatch([1, 2, 3], 1:3)
 cx = gpu(x)
-@test cx isa Flux.OneHotMatrix && cx isa CuArray
+@test cx isa Flux.OneHotMatrix && cx.data isa CuArray
```
Comment (thebhatman) [inline test/cuda/cuda.jl]: Can we do `cx.data` here? I think this would be an error.

Comment (dhairyagandhi96) [reply test/cuda/cuda.jl]: Why would it be an error?

Comment (MikeInnes) [reply test/cuda/cuda.jl]: `cx` is a OneHotMatrix, which is a wrapper around another array (called `data`). So yeah, it's fine, it's not the same as the `data` for a tracked array.

Comment (thebhatman) [reply test/cuda/cuda.jl]: Ooh okay! My bad, I thought `.data` was referring to the Tracker field data.

Comment (dhairyagandhi96) [reply test/cuda/cuda.jl]: Cheers!

File: test/cuda/cudnn.jl [hunk 5]
```diff
@@ -14,13 +14,11 @@ trainmode(f, x...) = forward(f, x...)[1]
...
-        g = rand(size(y)...)
-        # Flux.back!(y, g)
-        # Flux.back!(cy, gpu(g))
+        g = gradient(()->sum(m(x)), params(m))
+        cg = gradient(()->sum(cm(cx), params(cm))
 
-        @test m.γ ≈ cpu(cm.γ)
-        @test m.β ≈ cpu(cm.β)
-        @test x ≈ cpu(x)
+        @test g.grads[m.γ] ≈ cpu(cg.grads[cm.γ])
```
Comment (MikeInnes) [inline test/cuda/cudnn.jl]: You should index `g` directly here, rather than accessing `.grads`

File: src/layers/normalise.jl [hunk 6]
```diff
@@ -136,7 +136,7 @@ function (BN::BatchNorm)(x)
   dims = length(size(x))
   channels = size(x, dims-1)
   affine_shape = ntuple(i->i == ndims(x) - 1 ? size(x, i) : 1, ndims(x))
-  m = prod(size(x)[1:end-2]) * size(x)[end]
+  m = trunc(Int, prod(size(x))/channels)
```
Comment (MikeInnes) [inline src/layers/normalise.jl]: Use `÷` here.

Comment (thebhatman) [reply src/layers/normalise.jl]: That gives the same problem as `div`... I will just add `@nograd div` in zygote as you suggested..

Comment (MikeInnes) [reply src/layers/normalise.jl]: Yeah, that's fine. You can send a PR to Zygote and then update this code + the manifest.

Comment (thebhatman) [reply src/layers/normalise.jl]: So I change the manifest to depend on Zygote master, right? Do I just remove this line https://github.com/FluxML/Flux.jl/blob/b779d43aca84de06e0e9ff8618904d130eec2cbd/Manifest.toml#L344

Comment (MikeInnes) [reply src/layers/normalise.jl]: you shouldn't edit the manifest by hand. you can use the package manager to check the status (it's already on master, but and older commit), and just use `up` to get the latest master.

File: src/layers/normalise.jl [hunk 7]
```diff
@@ -136,7 +136,7 @@ function (BN::BatchNorm)(x)
   dims = length(size(x))
   channels = size(x, dims-1)
   affine_shape = ntuple(i->i == ndims(x) - 1 ? size(x, i) : 1, ndims(x))
-  m = trunc(Int, prod(size(x))/channels)
+  m = div(prod(size(x)), channels)
```
Comment (MikeInnes) [inline src/layers/normalise.jl]: Did you update the manifest, too?

Comment (thebhatman) [reply src/layers/normalise.jl]: Yes I did pkg up , but it didn't seem to have changed the Manifest. Just to verify, I did a fresh add of this branch and Zygote master, and it is working...

Comment (MikeInnes) [reply src/layers/normalise.jl]: You need to `activate` the Zygote project in order to use the package versions in the manifest, and also to affect it with `update`.

Comment (thebhatman) [reply src/layers/normalise.jl]: Thanks! I have updated the manifest now.

**Comment by thebhatman** [comment]
So once the adjoint for dropdims is merged https://github.com/FluxML/Zygote.jl/pull/264, backprop on InstanceNorm will work.. Only backprop on GroupNorm is failing due to this issue https://github.com/FluxML/Zygote.jl/issues/263

**Comment by dhairyagandhi96** [comment]
Very cool!

File: test/cuda/curnn.jl [hunk 8]
```diff
@@ -8,27 +8,27 @@ using Flux, CuArrays, Test
...
 
-      Δ = randn(size(y))
+      #Δ = randn(size(y))
 
-      Flux.back!(y, Δ)
-      Flux.back!(cuy, gpu(Δ))
+      #Flux.back!(y, Δ)
+      #Flux.back!(cuy, gpu(Δ))
 
-      @test x.grad ≈ collect(cux.grad)
```
Comment (MikeInnes) [inline test/cuda/curnn.jl]: The point of this code was to check gradients. Now it just checks that the cuda version of the model has the same parameters as the CPU version (and gradients don't work currently).

File: docs/src/training/optimisers.md [hunk 9]
```diff
@@ -3,25 +3,25 @@
 Consider a [simple linear regression](../models/basics.md). We create some dummy data, calculate a loss, and backpropagate to calculate gradients for the parameters `W` and `b`.
 
 ```julia
-using Flux, Flux.Tracker
+using Flux, Flux.Zygote
```
Comment (MikeInnes) [inline docs/src/training/optimisers.md]: @dhairyagandhi96 Flux already exports `gradient`, so this may not be necessary

File: docs/src/training/optimisers.md [hunk 10]
```diff
@@ -58,3 +58,78 @@ AMSGrad
 NADAM
 ADAMW
 ```
+
+## Optimiser Interface
```
Comment (MikeInnes) [inline docs/src/training/optimisers.md]: These new docs are great, but could we have them as a separate PR? Since they seem broadly independent of the Zygote changes.

File: docs/src/training/optimisers.md [hunk 11]
```diff
@@ -3,25 +3,25 @@
...
-predict(x) = W*x .+ b
+predict(x) = (W * x) .+ b
 loss(x, y) = sum((predict(x) .- y).^2)
 
 x, y = rand(5), rand(2) # Dummy data
 l = loss(x, y) # ~ 3
 
 θ = Params([W, b])
-grads = Tracker.gradient(() -> loss(x, y), θ)
+grads = Zygote.gradient(() -> loss(x, y), θ)
```
Comment (MikeInnes) [inline docs/src/training/optimisers.md]:
```suggestion
grads = gradient(() -> loss(x, y), θ)
```
@dhairyagandhi96

File: docs/src/training/optimisers.md (hunk 9 above)
Comment (dhairyagandhi96) [reply docs/src/training/optimisers.md]: Fixed

**Comment by MikeInnes** [comment]
Seem to be some issues with our GPU CI, so just merging.

File: src/layers/recurrent.jl [hunk 12]
```diff
@@ -42,21 +42,6 @@ end
...
-    truncate!(rnn)
-
-Truncates the gradient of the hidden state in recurrent layers. The value of the
-state is preserved. See also `reset!`.
-
-Assuming you have a `Recur` layer `rnn`, this is roughly equivalent to
-
-    rnn.state = Tracker.data(rnn.state)
-"""
-truncate!(m) = prefor(x -> x isa Recur && (x.state = _truncate(x.state)), m)
```
Comment (iblis17) [inline src/layers/recurrent.jl]: well, is there an alternative?
